File: Project/backend/app/utils/validation.py
"""Input validation utilities."""
import re
from typing import List


def validate_course_code(code: str) -> bool:
    """Validate course code format (e.g., 'CS 124', 'MATH 221')."""
    if not code or not isinstance(code, str):
        return False
    
    # Match pattern: 2-4 letters, space, 3 digits, optional letter
    pattern = r'^[A-Z]{2,4}\s+\d{3}[A-Z]?$'
    return bool(re.match(pattern, code.strip(), re.IGNORECASE))


def normalize_course_code(code: str) -> str:
    """Normalize course code to standard format."""
    if not code:
        return ''
    
    # Remove extra spaces and convert to uppercase
    normalized = code.strip().upper()
    
    # Ensure format is "DEPT 123" not "DEPT123"
    match = re.match(r'^([A-Z]{2,4})\s*(\d{3}[A-Z]?)$', normalized)
    if match:
        return f"{match.group(1)} {match.group(2)}"
    
    return normalized


def validate_course_codes(codes: List[str]) -> List[str]:
    """Validate and normalize list of course codes."""
    if not isinstance(codes, list):
        return []
    
    normalized = []
    for code in codes:
        if isinstance(code, str):
            norm_code = normalize_course_code(code)
            if norm_code and validate_course_code(norm_code):
                normalized.append(norm_code)
    
    return normalized

File: Project/Frontend/src/utils/validation.js
/**
 * Validation utilities for course codes and inputs
 */

/**
 * Validate course code format (e.g., "CS 124", "MATH 221")
 */
export const validateCourseCode = (code) => {
  if (!code || typeof code !== 'string') {
    return false;
  }
  
  // Match pattern: 2-4 letters, space, 3 digits, optional letter
  const pattern = /^[A-Z]{2,4}\s+\d{3}[A-Z]?$/i;
  return pattern.test(code.trim());
};

/**
 * Normalize course code to standard format
 */
export const normalizeCourseCode = (code) => {
  if (!code) return '';
  
  // Remove extra spaces and convert to uppercase
  const normalized = code.trim().toUpperCase();
  
  // Ensure format is "DEPT 123" not "DEPT123"
  const match = normalized.match(/^([A-Z]{2,4})\s*(\d{3}[A-Z]?)$/);
  if (match) {
    return `${match.group(1)} ${match.group(2)}`;
  }
  
  return normalized;
};

/**
 * Validate and normalize array of course codes
 */
export const validateAndNormalizeCourses = (courses) => {
  if (!Array.isArray(courses)) {
    return [];
  }
  
  return courses
    .map(code => normalizeCourseCode(code))
    .filter(code => code && validateCourseCode(code));
};

